#! /bin/sh /usr/share/dpatch/dpatch-run
## fix-module_admin.dpatch by Diego Iastrubni <diego.iastrubni@xorcom.com>
##
## DP: Create a connection to the manager in the module_admin
## DP: work around buggy freepbx-modules uninstall script (closes upstream 1731)
## DP: Howver, when trying to query for the version of asterisk and no manager
## DP: connection is available, ask asterisk directly via the console, about the
## DP: version (closes upstream 1899, using changest http://www.freepbx.org/trac/changeset/3572)

@DPATCH@
diff -urNad freepbx-2.2.1~dfsg~/amp_conf/bin/module_admin freepbx-2.2.1~dfsg/amp_conf/bin/module_admin
--- freepbx-2.2.1~dfsg~/amp_conf/bin/module_admin	2007-01-31 10:18:14.000000000 +0200
+++ freepbx-2.2.1~dfsg/amp_conf/bin/module_admin	2007-04-12 14:55:08.000000000 +0300
@@ -52,7 +52,7 @@
 }
 
 function init_amportal_environment($ampconfpath) {
-	global $amp_conf, $asterisk_conf, $db;
+	global $amp_conf, $asterisk_conf, $db, $astman;
 	
 	if (!file_exists($ampconfpath)) {
 		fatal('Cannot find conf file: '.$ampconfpath);
@@ -68,12 +68,18 @@
 	}	
 	// include the functions file from WEBROOT
 	include(AMP_BASE_INCLUDE_PATH.'/functions.inc.php');
-	
+	include(AMP_BASE_INCLUDE_PATH.'/common/php-asmanager.php');
+
 	// now apply the real parse function (this makes some default assumptions and does a bit more error checking)
 	$amp_conf = parse_amportal_conf($ampconfpath);
 	
 	$asterisk_conf = parse_asterisk_conf("/etc/asterisk/asterisk.conf");
 
+	$astman= new AGI_AsteriskManager();
+	if (! $res = $astman->connect("127.0.0.1", $amp_conf["AMPMGRUSER"] , $amp_conf["AMPMGRPASS"])) {
+		$astman = null;
+	}
+
 	// connect to database
 	if (!file_exists(AMP_BASE_INCLUDE_PATH.'/common/db_connect.php')) {
 		fatal('Cannot locate '.AMP_BASE_INCLUDE_PATH.'/common/db_connect.php');
@@ -464,7 +470,7 @@
 // freepbx modules will include the unsintall scripts
 // using relative paths. why not modifiying the include dir...?
 // fix for ticket:1731
-chdir ( $amp_conf\["AMPWEBROOT"\] . "/admin/" );
+chdir ( $amp_conf["AMPWEBROOT"] . "/admin/" );
 
 switch ($operation ) {
 	case 'list':
diff -urNad freepbx-2.2.1~dfsg~/amp_conf/htdocs/admin/functions.inc.php freepbx-2.2.1~dfsg/amp_conf/htdocs/admin/functions.inc.php
--- freepbx-2.2.1~dfsg~/amp_conf/htdocs/admin/functions.inc.php	2007-01-17 09:56:52.000000000 +0200
+++ freepbx-2.2.1~dfsg/amp_conf/htdocs/admin/functions.inc.php	2007-04-12 14:57:00.000000000 +0300
@@ -180,13 +180,13 @@
 
 	switch ($amp_conf['AMPENGINE']) {
 		case 'asterisk':
-			if ($astman) {
+			if (isset($astman) /*&& $astman->connected()*/) {
 				//get version
 				$response = $astman->send_request('Command', array('Command'=>'show version'));
 				$verinfo = $response['data'];
 			} else {
-				// could not connect to asterisk manager
-				return array('engine'=>'ERROR-UNABLE-TO-CONNECT', 'version'=>'0', 'additional' => '0');
+				// could not connect to asterisk manager, try console
+				$verinfo = exec('/usr/sbin/asterisk -V'); 
 			}
 			
 			if (preg_match('/Asterisk SVN.+/', $verinfo)) {
@@ -195,6 +195,8 @@
 			if (preg_match('/Asterisk (\d+(\.\d+)*)(-?(\S*))/', $verinfo, $matches)) {
 				return array('engine'=>'asterisk', 'version' => $matches[1], 'additional' => $matches[4]);
 			}
+
+			return array('engine'=>'ERROR-UNABLE-TO-CONNECT', 'version'=>'0', 'additional' => '0');
 		break;
 	}
 	return array('engine'=>'ERROR-UNSUPPORTED-ENGINE-'.$amp_conf['AMPENGINE'], 'version'=>'0', 'additional' => '0');
